flang/lib/Optimizer/Transforms/RewriteLoop.cpp

```diff
@@ -258,10 +258,23 @@ class CfgIterWhileConv : public mlir::OpRewritePattern<fir::IterWhileOp> {
 
     // With the body block done, we can fill in the condition block.
     rewriter.setInsertionPointToEnd(conditionBlock);
-    auto comp1 =
+    // The comparison depends on the sign of the step value. We fully expect
+    // this expression to be folded by the optimizer or LLVM. This expression
+    // is written this way so that `step == 0` always returns `false`.
+    auto zero = rewriter.create<mlir::ConstantIndexOp>(loc, 0);
```

Comment:
It may be rare for `step` to be unknown at compile time, but non-constant `upperbound`s won't be that rare.  Either of those will inhibit folding.

Reply:
This is not actually true. The logic is `((step < 0) && x) || ((step > 0) && y))` therefore if `step` is a constant, one of `(step < 0)`, `(step > 0)`, or both must be false and this folds to either `x`, `y`, or false. This can be seen by looking at some output.